src/bin/pg_autoctl/file_utils.c

```diff
@@ -262,6 +262,124 @@ read_file(const char *filePath, char **contents, long *fileSize)
 }
 
 
+/*
+ * move_file is a utility function to move a file from sourcePath to
+ * destinationPath. It behaves like mv system command. First attempts to move
+ * a file using rename. if it fails with EXDEV error, the function duplicates
+ * the source file with owner and permission information and removes it.
+ */
+bool
+move_file(char* sourcePath, char* destinationPath)
+{
+	if (strncmp(sourcePath, destinationPath, MAXPGPATH) == 0)
+	{
+		/* nothing to do */
+		log_warn("Source and destination are the same \"%s\", nothing to move.", sourcePath);
+		return true;
+	}
+
+	if (!file_exists(sourcePath))
+	{
+		log_error("Failed to move file, source file \"%s\" does not exist.", sourcePath);
+		return false;
+	}
+
+	if (file_exists(destinationPath))
+	{
+		log_error("Failed to move file, destination file \"%s\" already exists.", destinationPath);
+		return false;
+	}
+
+	/* first try atomic move operation */
+	if (rename(sourcePath, destinationPath) == 0)
+	{
+		return true;
+	}
+
+	/* rename fails with errno = EXDEV when moving file to a different file system */
+	if (errno != EXDEV)
+	{
+		log_error("Failed to move file \"%s\" to \"%s\": %s", sourcePath, destinationPath, strerror(errno));
+		return false;
+	}
+
+	if (!duplicate_file(sourcePath, destinationPath))
+	{
+		/* specific error is already logged */
+		log_error("Canceling file move due to errors.");
+		return false;
+	}
+
+	/* everything is successful we can remove the file */
+	unlink_file(sourcePath);
+
+	return true;
+}
+
+
+/*
+ * duplicate_file is a utility function to duplicate a file from sourcePath to
+ * destinationPath. It reads the contents of the source file and writes to the
+ * destination file. It expects non-existing destination file and does not
+ * copy over if it exists. The function returns true on successful execution.
+ *
+ * Note: the function reads the whole file into memory before copying out.
+ */
+bool
+duplicate_file(char* sourcePath, char* destinationPath)
+{
+	char *fileContents;
+	long fileSize;
+	struct stat sourceFileStat;
+	bool foundError = false;
+
+	if (!read_file(sourcePath, &fileContents, &fileSize))
+	{
+		/* errors are logged */
+		return false;
+	}
+
+	foundError = !write_file(fileContents, fileSize, destinationPath);
```

Comment:
This time we might be missing a `file_exists` test, that was placed in the previous function `move_file` where I don't think it belongs.

Reply:
added check for destination file